File: go/dsa/algorithms/union-find/union-find.go
type UnionFind struct { // @viz: def
    parent []int
    rank   []int
}

func NewUnionFind(n int) *UnionFind {
    parent := make([]int, n) // @viz: init
    rank := make([]int, n)
    for i := range parent { parent[i] = i }
    return &UnionFind{parent, rank}
}

func (uf *UnionFind) Find(i int) int {
    if uf.parent[i] == i { // @viz: f_check
        return i // @viz: f_return
    }
    // Path Compression
    uf.parent[i] = uf.Find(uf.parent[i]) // @viz: f_recurse
    return uf.parent[i] // @viz: f_return_node
}

func (uf *UnionFind) Union(i, j int) {
    rootI := uf.Find(i) // @viz: u_find_i
    rootJ := uf.Find(j) // @viz: u_find_j
    
    if rootI != rootJ { // @viz: u_check
        if uf.rank[rootI] > uf.rank[rootJ] { // @viz: u_rank_gt
            uf.parent[rootJ] = rootI // @viz: u_link_i
        } else if uf.rank[rootI] < uf.rank[rootJ] { // @viz: u_rank_lt
            uf.parent[rootI] = rootJ // @viz: u_link_j
        } else { // @viz: u_rank_eq
            uf.parent[rootJ] = rootI // @viz: u_link_same
            uf.rank[rootI]++ // @viz: u_rank_inc
        }
    }
}

File: python/dsa/algorithms/union-find/union-find.py
class UnionFind: # @viz: def
    def __init__(self, n):
        self.parent = list(range(n)) # @viz: init
        self.rank = [0] * n

    def find(self, i):
        if self.parent[i] == i: # @viz: f_check
            return i # @viz: f_return
        
        # Path Compression
        self.parent[i] = self.find(self.parent[i]) # @viz: f_recurse
        return self.parent[i] # @viz: f_return_node

    def union(self, i, j):
        root_i = self.find(i) # @viz: u_find_i
        root_j = self.find(j) # @viz: u_find_j
        
        if root_i != root_j: # @viz: u_check
            if self.rank[root_i] > self.rank[root_j]: # @viz: u_rank_gt
                self.parent[root_j] = root_i # @viz: u_link_i
            elif self.rank[root_i] < self.rank[root_j]: # @viz: u_rank_lt
                self.parent[root_i] = root_j # @viz: u_link_j
            else: # @viz: u_rank_eq
                self.parent[root_j] = root_i # @viz: u_link_same
                self.rank[root_i] += 1 # @viz: u_rank_inc